parlai/core/teachers.py

```diff
@@ -356,10 +359,16 @@ def next_episode_idx(self, num_eps=None, loop=None):
         if loop is None:
             loop = self.training
         if self.random:
-            new_idx = random.randrange(num_eps)
+            new_idx = (
+                random.randrange(
+                    num_eps // self.dws + int(num_eps % self.dws > self.rank)
```

Comment:
what is this computation exactly?

Reply:
we split across dws (distributed world size) workers.

if there are 1000 episodes and 16 workers, then 8 workers get 63 examples, and 8 workers get 62.

8 * 62 + 8 * 63 = 1000.

The > rank mess is identifying who gets the the 63 (the first workers) and who gets the 62 (the last workers)